### PR TITLE
Fixed / improved course description and relationship parsing

### PR DESCRIPTION
**Pull request by rstumbaugh**
Regex match at index 2 was always nil as there were only 2 matches: main match and 1 capture group which contained relevant info. Also added small formatting change for "Formerly" and "Also offered as".

### CODE REVIEW

**Comment by rstumbaugh** [comment]
Added some more changes regarding the course description and relationships. Now, if relationships are found, they are usually not included in the description. If relationships appear more than once, duplicates will appear in "additional_info" property of the course object. Also combined repeat fields (restrictions vs restricted to, credit granted vs credit only granted) into one field. 

I was worried that some of these changes might break apps that are already working around the old version. Is this a problem I should worry about or is it up to devs to keep an eye out for changes?

File: app/scrapers/courses_scraper.rb [hunk 1]
```diff
@@ -53,45 +53,117 @@ def utf_safe text
...
+  puts "Getting courses for #{dept_id} (#{semester})"
+
+  page = Nokogiri::HTML(open(url), nil, "UTF-8")
+
   department = page.search('span.course-prefix-name').text.strip
 
-  page.search('div.course').each do |course|
-    description =  (utf_safe(course.css('div.approved-course-texts-container').text + course.css('div.course-texts-container').text)).strip.gsub(/\t|\r\n/,'')
+
+  page.search('div.course').each { |course|
```
Comment (nickaversano) [inline app/scrapers/courses_scraper.rb]: nit: ruby way of doing multiline blocks is `.each do |course|`

File: app/scrapers/courses_scraper.rb [hunk 2]
```diff
@@ -53,45 +53,117 @@ def utf_safe text
...
-  end
+  }
   
-  puts "inserting courses from #{dept_id} (#{semester})"
-  # bulk for speed, upsert so it can be run multiple times without worry
-  course_array.each do |course|
+
+  courses.each { |course|
     bulk.find({course_id: course[:course_id]}).upsert.update({ "$set" => course })
-  end
```
Comment (nickaversano) [inline app/scrapers/courses_scraper.rb]: keep this (to close do block)

File: app/scrapers/courses_scraper.rb [hunk 3]
```diff
@@ -53,45 +53,117 @@ def utf_safe text
...
     }
-
-  end
+  }
   
-  puts "inserting courses from #{dept_id} (#{semester})"
-  # bulk for speed, upsert so it can be run multiple times without worry
-  course_array.each do |course|
+
+  courses.each { |course|
```
Comment (nickaversano) [inline app/scrapers/courses_scraper.rb]: same

File: app/scrapers/courses_scraper.rb [hunk 4]
```diff
@@ -53,45 +53,117 @@ def utf_safe text
...
+    courses << {
+      course_id: course_id,
+      name: course_title,
       dept_id: dept_id,
       department: department,
       semester: semester,
       credits: course.css('span.course-min-credits').first.content,
-      grading_method: course.at_css('span.grading-method abbr') ? course.at_css('span.grading-method abbr').attr('title').split(', ') : [],
+      grading_method: course.at_css('span.grading-method abbr') ? 
+              course.at_css('span.grading-method abbr').attr('title').split(', ') : [],
```
Comment (nickaversano) [inline app/scrapers/courses_scraper.rb]: nit: safe navigation operator (`&`) could also be used here, e.g.: `course.at_css('span.grading-method abbr')&.attr('title')&.split(', ') || []`
Where anything that evals to nil will stop at the first ampersand and the whole expression becomes nil. Also using `nil || []` evaluates to `[]`

File: app/scrapers/courses_scraper.rb [hunk 5]
```diff
@@ -53,45 +53,117 @@ def utf_safe text
...
+    match = /Formerly:? ([^.]+\.)/.match(text)
+    text = match ? text.gsub(match[0], '') : text
+    formerly = match ? match[1] : nil
+
+    match = /Additional information: ([^.]+\.)/.match(text)
+    text = match ? text.gsub(match[0], '') : text
+    additional_info = match ? match[1] : nil
+
+    # if approved-course-texts held relationships, use 2nd child as description and leftover text as "additional info"
+    if approved.css('> div').length > 0 then
```
Comment (nickaversano) [inline app/scrapers/courses_scraper.rb]: nit: don't use `then`

File: app/scrapers/courses_scraper.rb [hunk 6]
```diff
@@ -53,45 +53,117 @@ def utf_safe text
...
+
+      description = utf_safe approved.css('> div:last-child').text.strip.gsub(/\t|(\r\n)/, '')
+      additional_info = additional_info ? additional_info += ' '+text : text
+      additional_info = additional_info && additional_info.strip.empty? ? nil : additional_info.strip
+
+    elsif other.css('> div').length > 0 then
```
Comment (nickaversano) [inline app/scrapers/courses_scraper.rb]: `then`

File: app/scrapers/courses_scraper.rb [hunk 7]
```diff
@@ -53,45 +53,117 @@ def utf_safe text
...
+    #
+    # algorithm finds relationships and, if they exist, removes them from the description
+    # searches both approved-course-texts > div:first-child and course-texts > div for relationships
+    # leftover text will either be description (if approved-course-texts is empty) or additional info
+
+    approved = course.search('div.approved-course-texts-container')
+    other = course.search('div.course-texts-container')
+
+    # get all relationship text
+    if approved.css('> div').length > 1 then
```
Comment (nickaversano) [inline app/scrapers/courses_scraper.rb]: don't use then. it's optional and not the ruby-way

**Comment by rstumbaugh** [comment]
Hey Nick, glad you like the changes and thanks for the comments! I've never used Ruby outside of the classroom, so I appreciate any comments on convention / general style. The scraper works fine on my end so it should be good to go!

**Comment by rstumbaugh** [comment]
It's looking like the safe navigation operator is a ruby 2.3 feature. I updated ruby on my Vagrant box but wanted to make sure it was updated on the server.

**Pull request merged**